File: Exercises/R/machine-learning-ex1/computeCost.R
df <- read.table("/Users/cmr/Desktop/Machine Learning/Class/Octave/machine-learning-ex1/ex1/ex1data1.txt", header = FALSE, sep = ",")
colnames(df)[1] <- "Population"
colnames(df)[2] <- "Profit"
#matrix(0, n, m) #creates a matrix with n lines and m columns, filled with zeros.

X <- cbind(a=1, df$Population);
y<-df[,c("Profit")];
m<-length(df$Profit);
theta<-matrix(0,2,1);

computeCost <- function(matrixA, matrixB, theta_matrix, m){
	predictions<-matrixA%*%theta_matrix;
	squaredErrors<-(predictions-matrixB)^2;
	J=(1/(2*m))*sum(squaredErrors);
	return (J);
}

new_theta <- matrix(c(-1, 2), ncol = 1);

cat("With theta=[0;0] cost computed= ", computeCost(X, y, theta, m));
print("Expected cost value (approx) 32.07");

cat("With theta=[0;0] cost computed= ", computeCost(X, y, new_theta, m));
print("Expected cost value (approx) 54.24");

File: Exercises/Python/machine-learning-ex1/computeCost.py
from scipy import linalg, sparse #for linalg
import matplotlib.pyplot as plt #for the graph
import numpy #for lin alg operations
import pandas as pd #for reading in file


def computeCost(matrix, theta_value):
	m=matrix.shape[0]; #gets the number of training ex
	X=numpy.asmatrix(matrix.values[:,0]).T; #zero indxed columns
	y=numpy.asmatrix(matrix.values[:,1]).T; #real values
	ones = numpy.ones((m, 1)); #column of 1s
	X=numpy.hstack((ones,X)); #concatenate 1s to represent X0
	predictions=numpy.dot(X, theta_value);
	squaredError=numpy.square((predictions-y));
	sumValue = squaredError.sum(axis=0).item(0); #sum() returns a matrix, so get the item and store it
	return (1./(2*m))*sumValue;


data = pd.read_csv('ex1data1.txt', header=None, delimiter=','); #data is a 97x2 matrix
theta = numpy.zeros((2, 1)); #fitting parameters
print(computeCost(data, theta));
print("Expected cost value (approx) 32.07\n");

newTheta=numpy.mat([[-1], [2]]);
print(computeCost(data, newTheta));
print("Expected cost value (approx) 54.24\n");
